### ISSUE TITLE
DagNode.addChild crashes in certain circumstances

### ISSUE THREAD

**Issue by wougzy** [mottosso/cmdx#28]
i identified a crash with addChild easy to reproduce

try this :
```
import cmdx as mx
j0 = mx.create_node(mx.tJoint)
j1 = mx.create_node(mx.tJoint, parent=j0)
j2 = mx.create_node(mx.tJoint, parent=j1)
j0.add_child(j2)
```

it crashes probably because j2 is a descendent of j0. you can fix it by removing the hierarchy link before like this

```
    def addChild(self, child, index=Last):
        mobject = child._mobject
        parent = child.parent()
        if parent is not None:
            parent._fn.removeChild(mobject)
        self._fn.addChild(mobject, index)
```
i think this is the simplest form possible since checking ancestors would require to find parents anyway. but i'm not sure about the performance impact of removing children if not necessary
what do you think?

**Comment by mottosso** [comment]
Thanks for reporting this. I'm glad you consider the performance impact of your suggestions. :)

What I would suggest is that you profile the change, before and after, identify what the cost is and if it's "too much" then we could find a way to work with that; e.g. either make it optional, or add a stern warning in the documentation.

To profile things like this, have a look at [plot.py](https://github.com/mottosso/cmdx/blob/master/plot.py) for inspiration.

**Comment by wougzy** [comment]
looks like it's safe. i ran several loops of 20000 parent commands and the difference between with or without is notable but still very small (about 50ms vs 200ms for each packet)

**Comment by mottosso** [comment]
If you can fit your test in `plot.py`, ideally using a similar/same convention, then that would be swell. Then you could make a note of how long it takes ong that line, for anyone looking at this in the future wondering how costly that is. You'll find similar notes across the code, in a `ns` unit.

**Comment by wougzy** [comment]
hmm ok i see! here are more serious results (: is that what you had in mind?

```
New()

root1 = cmdx.createNode("transform")
root2 = cmdx.createNode("transform")
child = cmdx.createNode("transform", parent=root2)

melparent = 'parent -r "transform3" "transform1"; parent -r "transform3" "transform2";'
Test("mel", "parent", lambda: mel.eval(melparent), number=1, repeat=1000)
Test("cmds", "parent", lambda: [cmds.parent("transform3", parent, r=1) for parent in ("transform1", "transform2")], number=1, repeat=1000)
Test("cmdx", "parent", lambda: [cmdx.parent(child, parent) for parent in (root1, root2)], number=1, repeat=1000)
Test("cmdx safe", "parent", lambda: [cmdx.parent(child, parent, safe=True) for parent in (root1, root2)], number=1, repeat=1000)
root1, root2, child = pm.ls(map(str, (root1, root2, child)))
Test("PyMEL", "parent", lambda: [pm.parent(child, parent, r=1) for parent in (root1, root2)], number=1, repeat=1000)
```

```
parent mel: 115.5 ms (109.30 µs/call)
parent cmds: 119.1 ms (114.50 µs/call)
parent cmdx: 11.1 ms (10.60 µs/call)
parent cmdx safe: 19.3 ms (18.30 µs/call)
parent PyMEL: 601.7 ms (583.70 µs/call)
```

it looks like adding this security costs about twice the original command. still way cheaper than mel/cmds though

**Comment by mottosso** [comment]
Oh you are a star. :D Thanks for putting a smile on my face.

Yes, that is exactly what I had in mind. That `safe=` option is great, we could leave that in, and default it to `True`. That way, anyone wanting to squeeze performance could toggle that flag, and they'd be good to go.

**Comment by mottosso** [comment]
Implemented in #24. Thanks @wougzy!
